### ISSUE TITLE
Righteous Fury has an incorrect threat modifier + some holy spells aren't being accounted for.

### ISSUE THREAD

**Issue by Hark47** [dfherr/LibThreatClassic2#7]
Here I posted this in the paladin discord, I'll just paste it here

"the more I'm looking at it the more it seems that all RF calculations are wrong on tiny threat
judgment of rightouesness hit for 278 but tiny threat only shows 414 threat which is less than it should be
hitting exorcisms for 587 but showing as 1.4k threat which is way too much

Okay figured it out
It incorrectly has the RF threat modifier at 2.4 and some spells aren't being accounted for
The reason it shows 414 threat from Judgement of Rigtheousness is cause it shows all judgements as having a base threat of 57 (which could be true for the debuffs but likely doesn't apply to JoR), when modified with the incorrect 2.4 that's 136.8 threat, and for some reason tiny threat isn't scaling the threat from JoR with RF and leaving it at it's base threat. So 278 + 136  = 414
60 threat from GBoK x 2.4 = 144
587 damage from Exorcism x 2.4 = 1.4k"

**Comment by dfherr** [comment]
Hi thanks for reporting a bug.

Unfortunately your comment is a little bit confusing. What threat from JoR? Its damage? Which number is the damage? I don't have a paladin myself. These are the id's tracked for JoR
```
-- Judgement of Righteousness
	20187, 20280, 20281, 20282, 20283, 20284, 20285, 20286,
```

Can you turn on out-of-game combat `/combatlog` and paste me the combat log entry for Judgment of Righteousness? Also, maybe spell_id lookup doesn't work for Judgments, because `GetSpellInfo()` doesn't work for judgments.

Why is 2.4 for RF threat wrong? 60% base and 50% from talent multiply to 140%. Is this talent additive instead of multiplicative like other talents?

**Comment by Hark47** [comment]
With Improved Righteous Fury the threat modifier for almost all holy spells is 1.9, the talent saying "increases by 50%" it means it's going from 60% increase to 90% threat increase. The only holy spells that don't have a modifier of 1.9 are our heals (.25 x 1.9 = modifier of .475) and mana regen from Seal of Wisdom or Judgement of Wisdom (.5 x 1.9 = modifier of .95)

And when I judge righteousness the only line that mentions the spell ID for JoR is this one

> 1/3 15:15:41.619  SPELL_DAMAGE,Player-4408-000D6E10,"Hark-Faerlina",0x511,0x0,Creature-0-4411-1-226-7443-00000FA711,"Shardtooth Mauler",0x10a48,0x0,20286,"Judgement of Righteousness",0x2,Creature-0-4411-1-226-7443-00000FA711,0000000000000000,94,100,0,0,0,-1,0,0,0,6858.89,-4864.28,1452,0.8485,56,277,277,-1,2,0,0,0,nil,nil,nil

The full cast of a JoR is this

`1/3 15:15:37.429  SPELL_AURA_APPLIED,Player-4408-000D6E10,"Hark-Faerlina",0x511,0x0,Player-4408-000D6E10,"Hark-Faerlina",0x511,0x0,20293,"Seal of Righteousness",0x2,BUFF`

`1/3 15:15:37.429  SPELL_CAST_SUCCESS,Player-4408-000D6E10,"Hark-Faerlina",0x511,0x0,0000000000000000,nil,0x80000000,0x80000000,20293,"Seal of Righteousness",0x2,Player-4408-000D6E10,0000000000000000,91,100,0,0,0,-1,0,0,0,6858.70,-4842.46,1452,4.9396,62`

`1/3 15:15:40.928  SPELL_AURA_REMOVED,Player-4408-000D6E10,"Hark-Faerlina",0x511,0x0,Player-4408-000D6E10,"Hark-Faerlina",0x511,0x0,20293,"Seal of Righteousness",0x2,BUFF`

`1/3 15:15:40.928  SPELL_CAST_SUCCESS,Player-4408-000D6E10,"Hark-Faerlina",0x511,0x0,Creature-0-4411-1-226-7443-00000FA711,"Shardtooth Mauler",0x10a48,0x0,20271,"Judgement",0x2,Player-4408-000D6E10,0000000000000000,93,100,0,0,0,-1,0,0,0,6861.10,-4861.77,1452,0.9125,62`

`1/3 15:15:41.619  SPELL_DAMAGE,Player-4408-000D6E10,"Hark-Faerlina",0x511,0x0,Creature-0-4411-1-226-7443-00000FA711,"Shardtooth Mauler",0x10a48,0x0,20286,"Judgement of Righteousness",0x2,Creature-0-4411-1-226-7443-00000FA711,0000000000000000,94,100,0,0,0,-1,0,0,0,6858.89,-4864.28,1452,0.8485,56,277,277,-1,2,0,0,0,nil,nil,nil`

**Comment by dfherr** [comment]
Thanks for the info. I guess `GetSpellInfo(20286)` returns nil for Paladins, because `Judgment of Righteousness` is not in your spellbook, but just the Seal and the spell Judgement. Because WoW classic completely removed spell_ids from the ingame combatlog, I need to handle JoR directly based on name. You could verify that ingame by running `/run print(GetSpellInfo(20286))`. If it returns `nil` instead of the spell information, this is the source of that bug.

So Paladins have a 0.25 heal threat modifier instead of the "known" 0.5?

Mana regen from Seal of Wisdom gets threat from RF? Does this count as a holy attack!?

**Comment by Hark47** [comment]
When running that script I see "Judgement of Rightousness nil 132325 0 0 100 20286"

Yeah, all paladin heals have a 0.25 modifier to prevent them from using Improved Righteous Fury for OP healing threat

All unnatural mana gains (like those from abilities or potions) generate threat on a 0.5 threat modifier, this applies to mana gained from SoW/JoW. It also generates threat on all mobs you currently have threat on, like if a mage is AoEing a pack and pops a potion that gives 1.5k mana he'll generate 750 threat on every single target. Mana regen from MP5 and spirit do no generate threat though.

And yeah Righteous Fury gives a 90% threat increase to all holy effects even though the description just says "holy attacks".

**Comment by dfherr** [comment]
> And yeah Righteous Fury gives a 90% threat increase to all holy effects even though the description just says "holy attacks".

That I know. But mana gain is certainly not a holy effect, so I find it weird that mana gain from seal of wisdom should be affected by that.

A mana potion generates 750 threat, but that threat should be divided evenly among all targets similar to healing threat. So a mage fighting 5 mobs and using a mana potion adds 150 threat to each mob, not 750. At least that's what I found in sources online.

Is your information all backed up by ingame tests in the current client? Is there a source for it I can verify and link to when implementing the changes?

**Comment by Hark47** [comment]
Heal threat being split was a bug and has been patched out, it seems: https://us.forums.blizzard.com/en/wow/t/demoralizing-shout-is-bugged-and-it-is-why-we-have-this-aoe-meta/295678/64

And as for Seal of Wisdom / Judgement of Wisdom, the mana gain is a holy effect. Mana gains are coded like heals and can be in a certain school of magic. Mana pots are physical effects, life tap is shadow (but was coded to give no threat), JoW and SoW are holy effects. I can say that the AoE threat of JoW/SoW is VERY noticeable when tanking dungeons.

**Comment by Hark47** [comment]
Did not mean to close, I'm tired as hell lol

**Comment by dfherr** [comment]
I will fix this with the next revision, but I need to find time and a friendly Paladin to help me test things out, so I don't release another buggy version

**Comment by mspk7305** [comment]
I'll test whenever you're ready

**Comment by Hark47** [comment]
yeah I can help test whenever

**Comment by dfherr** [comment]
I fixed the righteous fury modifier to `1 + 0.6 * talent`, so it should be 90% bonus threat with the talent now.

I also changed the whole righteous fury handling, so it is simply checking for the spell school being holy. So everything considered holy now gets bonus threat from righteous fury.

I also checked and all paladin heals correctly get an additional 50% threat reduce (so 25% for heals by default).

Furthermore I added the spellschool handler to mana gains, so holy mana gains with righteous fury should get bonus threat.

I created an alpha release for testing: https://github.com/dfherr/LibThreatClassic2/releases/tag/r6-alpha1
Replace one of your LibThreatClassic2 installations with the contents of the alpha release. Debugging is turned on and will print ADD_THREAT statements to your fourth chat window (create it and reload if you don't have one).
You don't have to replace all installations, just one of them should work. In fact, if it doesn't work, it's a major bug so reporting would be appreciated ;)

**Comment by MysticalOS** [comment]
Eww, I tagged a new mod release. Some proejects just auto pull the latest version of lib. so I guess people will deal with threat debug messages for the next month being build into DBM :D 

A tip for debugging in future. just add a debug mode that exists in all versions, even release. But one that can be toggled on and off by users via some option variable that you can toggle via script or slash. :)

**Comment by dfherr** [comment]
Luckily I was smart enough to not commit the enabled debug mode. It's just in the downloadable zip file :)

so people should be fine, but revision 6 will never get a non-alpha release. The next release will be revision 7 then.

**Comment by dfherr** [comment]
@Silverthane2864 @mspk7305 did you get a chance to verify if the changes have the supposed effects? Would be good to know before I start releasing versions that build on top of these changes. ThreatClassic2 also provides these changes as an alpha version installable through twitch.

**Comment by mspk7305** [comment]
Tonight is my first night back to raiding, will know in a couple hours.

**Comment by Hark47** [comment]
From my testing everything seems great with these changes, I look forward to seeing this soon :D

**Comment by mspk7305** [comment]
I was unable to get the described debug log output but I suspect it is due
to DBM conflicting with Details

**Comment by dfherr** [comment]
> 
> I was unable to get the described debug log output but I suspect it is due to DBM conflicting with Details

@mspk7305 no this should not be a conflict. The version has a higher version number, if this fails because of a conflict something is seriously broken, so I really hope you just did a mistake.

Did you download `LibThreatClassic2-r6-alpha1.zip` from here: https://github.com/dfherr/LibThreatClassic2/releases/tag/r6-alpha1 That's the ONLY one that has debugging enabled.

Did you correctly replace one LibThreatClassic2 installation in your wow folders?

Do you have 4th chat windows ingame? (your 2nd chat window is usually your combat log). Debug output is only on the 4th window.

As a side note, debug output is really not supposed to be used in a raid. It will just spam your chat. It's great for testing alone or in a party of two though.

**Comment by mspk7305** [comment]
Can confirm LibThreatClassic2-r6-alpha1, and the install location:
[image: image.png]
Can also confirm 4 chat windows though my 4th one is setup to only show
loot messages. Not sure this prevented messages from appearing or if there
may be something else wonky with addons here.

**Comment by dfherr** [comment]
Oh, yea, I just remembered that DBM did pull in version 6 in the new releases. so if you have an updated DBM and don't replace it in DBM, you might load the DBM version without debug enabled.

**Comment by mspk7305** [comment]
That's likely it then, I'll try without dbm

**Comment by BigRaj** [comment]
My dedicated paladin tank and I spent about a day dedicated to testing paladin threat.  

`return self:RighteousFury(amt) * 1.2` on line Line 138  is incorrect because line 116 `self.schoolThreatMods[SCHOOL_MASK_HOLY] = self.RighteousFury`  already applies the RighteousFury mod.  making Righteous fury apply twice on the same spell and over reporting threat.  

Line 138 was changed to `return amt * 1.2` and that fixed it.  

Second issue we noticed was that Judgement did not actually cause 57 threat.  Casting a judgement with no damage reported 57 threat, however I was able to pull from range with a lvl 1 wand in 1 hit.  

Our solution was to change it to 0.5 threat per judgement, however that is not a known value for us.  

Our updated Paladin.lua can be found here.  
https://gist.github.com/BigRaj/6bea87189e8a27c39b558ed0c9fbdeef
